cdap-hbase-compat-0.96/src/main/java/co/cask/cdap/data2/increment/hbase96/IncrementHandler.java

```diff
@@ -120,6 +124,67 @@ public void preGetOp(ObserverContext<RegionCoprocessorEnvironment> ctx, Get get,
   }
 
   @Override
+  public Result preIncrement(ObserverContext<RegionCoprocessorEnvironment> e, Increment increment)
+    throws IOException {
+    // note that in 0.96, there is no hook for increment after the region lock was obtained
+    // therefore there is a possibility of race conditions when using this HBase version.
+
+    boolean isIncrement = increment.getAttribute(HBaseTable.DELTA_WRITE) != null;
+    if (!isIncrement) {
+      return null; // this should only trigger for a readless increment
+    }
+    boolean transactional = state.containsTransactionalFamily(increment.getFamilyCellMap().keySet());
+    byte[] txBytes = increment.getAttribute(TxConstants.TX_OPERATION_ATTRIBUTE_KEY);
+    if (transactional && txBytes == null) {
+      throw new IllegalArgumentException("Attribute " + TxConstants.TX_OPERATION_ATTRIBUTE_KEY
+                                           + " must be set for transactional readless increments");
+    }
+    long tsToAssign;
+    if (transactional) {
```

Comment:
If it is not transactional, isn't that we can just turn the `Increment` into a `Put` and call `region.Put` (with bypass as well) without doing the rest of the logic? The `Put` should trigger the `prePut` logic and do the right thing?